### ISSUE TITLE
<ul type> and <ol type> not supported by List nor ListProperties

### ISSUE THREAD

**Issue by wimleers** [ckeditor/ckeditor5#11615]
## 📝 Provide detailed reproduction steps (if any)

While bringing the `<ol start>`  and `<ol reversed>`  functionality to Drupal core, we’re also looking at `<ol type>` and `<ul type>`, because as https://ckeditor.com/blog/ckeditor-5-v32.0.0-with-new-list-properties-support-for-the-script-tag-and-enhanced-mentions/ mentions, this is now supported too and has a magnificent UX. 

https://ckeditor.com/docs/ckeditor5/latest/api/module_list_listproperties-ListPropertiesConfig.html#member-styles says:
> When set, the list style feature will be enabled. It allows changing the list-style-type HTML attribute of the lists.

But:
1. There is no `list-style-type` HTML attribute. Only a `type` attribute.
2. There seems to be no way to configure CKEditor 5 to generate a type attribute, only `list-style-type`.
3. Searching the CKE5 codebase finds plenty of matches for the strings `<ul style="list-style-type` and `<ol style="list-style-type`, but zero for `<ul type` and `<ol type`

**Conclusion:**

Because `ListProperties`' support for list style types does _not_ work for `<ol type>` or `<ul type>`, despite that being in the HTML spec:

- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ol#attr-type → not deprecated
- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ul#attr-type → deprecated _but supported_

→ No downcast support, but also, more worryingly still, no upcast support.

Therefore the only work-around is to use GHS to add support for these. But … _that also does not work_:

1. Load the following CKE5 plugins: `htmlSupport.GeneralHtmlSupport`, `list.List`, with the following `htmlSupport` config:
```
{"allow":[{"name":"ul","attributes":{"type":true},"classes":true,"styles":true}]}
```
2. Instantiate CKEditor 5 on a textarea containing markup like:
```
<ol type="A">
  <li>test</li>
</ol>
```

**I strongly suspect that `v32` caused this to not work with GHS anymore, because I do recall testing this a long time ago and it worked fine then.**

### ✔️ Expected result
CKEditor 5 retains the `type` attribute.

### ❌ Actual result
It does not.

## 📃 Other details

* Browser: N/A — reproduced in Chrome and Firefox.
* OS: N/A — reproduced in macOS Big Sur
* First affected CKEditor version: `v33.0.0`
* Installed CKEditor plugins: see above.

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

**Comment by wimleers** [comment]
Sibling Drupal issue: https://www.drupal.org/project/drupal/issues/3274651

**Comment by wimleers** [comment]
Related: #11595.

**Comment by Reinmar** [comment]
Quick note about GHS + list: The old list feature had no integration with GHS. Only the new one (document lists) has. 

> **I strongly suspect that `v32` caused this to not work with GHS anymore, because I do recall testing this a long time ago and it worked fine then.**

AFAIR it didn't work when we discussed it the last time. Also, I think it could never work via GHS because we never had integration with lists for it.

**Comment by Reinmar** [comment]
I've just checked that loading the below content to CKEditor 4:

```
<ol type="A">
	<li>sdfsdf</li>
	<li>sdfsdf</li>
	<li>sfsdf</li>
</ol>
```

Makes it return this data:

```
<ol style="list-style-type:upper-alpha">
	<li>sdfsdf</li>
	<li>sdfsdf</li>
	<li>sfsdf</li>
</ol>
```

In other words, CKEditor 4 normalizes `type` to `list-style-type`.

We could add similar logic to the CKEditor 5 document list's upcast.

**Comment by wimleers** [comment]
> Quick note about GHS + list: The old list feature had no integration with GHS. Only the new one (document lists) has. 

Does that mean that per this bit of https://github.com/ckeditor/ckeditor5/issues/2973#issuecomment-1089355036:

> This feature will be a part of the upcoming v34.0.0 release. However, due to its size and complexity, we'll keep it on an "experimental" level for 1-2 releases, meaning that we don't recommend using it in production without assessing the risk. For now, it isn't even covered in the documentation, although this should change sooner.

… Drupal updating to CKEditor 5 `v34.0.0` didn't actually solve this for us?

**Comment by lauriii** [comment]
I did some research on whether it would be feasible to override the downcast and upcast for this per request from @wimleers.

We could do a pretty scoped override if we could either override `createAttributeStrategies` or the array returned by that function. Is that something that CKEditor 5 could provide us as an API for this use case?

If that's not feasible, we could still override the upcast and downcast for the `listStyle` attribute and hardcode the assumptions from `list/listproperties/listpropertiesediting` in our code. What could be helpful is if you could provide at least `areRepresentingSameList` function as an exported function so we wouldn't have to duplicate that, since it seems like something that would change in case the list plugin receives new features.

**Comment by Reinmar** [comment]
We discussed this on a refinement meeting whether it'll be easier for us to make the current downcast/upcast strategies easily extensible/overridable or simply implementing the support for `config.list.properties.styles.useAttribute = true`.

We came to a conclusion that the config option will be easier.

**Comment by Reinmar** [comment]
Scope:

* The changes will affect only `DocumentList`.
* Extend the upcast callback for the `type` attribute. In case when both the attribute and style are specified, follow what the browser prioritizes.
* Add support for the `config.list.properties.styles.useAttribute = true` option.
* Document this in the API docs for `ListPropertiesConfig`. Explain that it works only in `DocumentListProperties`.
* Make sure to map the values on upcast and downcast. We should use the same set of values in the model, regardless of whether a style or an attribute is used in the data.

**Comment by Reinmar** [comment]
There's a problem that the \`type\` attribute supports a subset of the options supported by CSS and by our UI.

Values supported by the attribute:

> After https://html.spec.whatwg.org/multipage/grouping-content.html#attr-ol-type
> 
> | Keyword | State | Description | Examples for values 1-3 and 3999-4001 |
> | --- | --- | --- | --- |
> | `1` (U+0031) | decimal | Decimal numbers | 1. | 2. | 3. | ... | 3999. | 4000. | 4001. | ... |
> | `a` (U+0061) | lower-alpha | Lowercase latin alphabet | a. | b. | c. | ... | ewu. | ewv. | eww. | ... |
> | `A` (U+0041) | upper-alpha | Uppercase latin alphabet | A. | B. | C. | ... | EWU. | EWV. | EWW. | ... |
> | `i` (U+0069) | lower-roman | Lowercase roman numerals | i. | ii. | iii. | ... | mmmcmxcix. | i̅v̅. | i̅v̅i. | ... |
> | `I` (U+0049) | upper-roman | Uppercase roman numerals | I. | II. | III. | ... | MMMCMXCIX. | I̅V̅. | I̅V̅I. | ... |

Our UI:

![](https://user-images.githubusercontent.com/156149/165487965-99ef89e4-f047-4af6-a3e0-a7673757b2a9.png)

So the attribute does not support the `01.` option.

Proposed approach:

*   Hide the `01.` option in the UI when the `useAttribute` option is used.
*   When upcasting from `list-style-type`, when we have a value unsupported by the `type` attribute, use `DEFAULT_LIST_TYPE`. In other words, when `useAttribute` is `true` we may have only 5 values in the model.

**Comment by wimleers** [comment]
> *and* not retained in GHS #11615

This piece of the title can be removed — see https://github.com/ckeditor/ckeditor5/issues/11595#issuecomment-1112136660 😊 


> Scope:
> 
> * The changes will affect only `DocumentList`.
> * Extend the upcast callback for the `type` attribute. In case when both the attribute and style are specified, follow what the browser prioritizes.
> * Add support for the `config.list.properties.styles.useAttribute = true` option.
> * Document this in the API docs for `ListPropertiesConfig`. Explain that it works only in `DocumentListProperties`.
> * Make sure to map the values on upcast and downcast. We should use the same set of values in the model, regardless of whether a style or an attribute is used in the data.

👍  Sounds like a great plan!




> There's a problem that the `type` attribute supports a subset of the options supported by CSS and by our UI.
> […] 
> Proposed approach:
> 
> * Hide the `01.` option in the UI when the `useAttribute` option is used.
> * When upcasting from `list-style-type`, when we have a value unsupported by the `type` attribute, use `DEFAULT_LIST_TYPE`. In other words, when `useAttribute` is `true` we may have only 5 values in the model.

👍 This seems totally reasonable 😊

**Comment by wimleers** [comment]
Just to make sure we’re on the same page: because the “hard crash” part of `<ol type>` and `<ul type>` was indeed fixed by Drupal updating to https://www.npmjs.com/package/@ckeditor/ckeditor5-list at `34.0.1` (which you released 3 days ago), #11615 (this issue) is not critical.

IOW: your theory was correct, once we switched to `DocumentList`, supporting `<ul type>` and `<ol type>` through GHS works fine 👍  We still want #11615 (this issue) to be able to offer your vastly superior native UX though!

For that we have https://www.drupal.org/project/drupal/issues/3274635. 🚀
